File: python/src/coasty/cost.py
"""Pure, offline cost estimator implementing the full Coasty pricing table.

1 credit = 1 cent = $0.01 exactly. Sandbox keys (``sk-coasty-test-*``) never
bill -- pass ``sandbox=True`` to :func:`format_estimate` to label that.

Pricing (per docs/API_NOTES.md):

- predict 5 cr, session create 10 cr (no surcharges), session step 4 cr,
  ground 3 cr (+1 if HD), parse 0 cr.
- Surcharges (predict + session step): +2 cr per trajectory screenshot,
  +1 cr per HD image (strictly w>1280 OR h>720 -- exactly 1280x720 is NOT
  HD; applies to the current AND each trajectory screenshot), +3 cr on the
  v1 engine, +1 cr when system_prompt exceeds 500 chars (exactly 500 free).
- Run step: 5 cr on v3/v4, 8 cr on v1 (no other surcharges). Workflow task
  steps bill like run steps; control-flow steps are free.
- Machines (hourly, metered per minute, rounded down): Linux running 5/hr,
  Windows running 9/hr (incl. starting/stopping/restarting), stopped or
  suspended 1/hr any OS, creating/error/terminated free. Snapshot 1 cr.
"""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass

from .types import CuaVersion, OsType

CREDIT_USD = 0.01

PREDICT_BASE_CREDITS = 5
SESSION_CREATE_CREDITS = 10
SESSION_PREDICT_BASE_CREDITS = 4
GROUND_BASE_CREDITS = 3
PARSE_CREDITS = 0
TRAJECTORY_SURCHARGE_CREDITS = 2
HD_SURCHARGE_CREDITS = 1
V1_SURCHARGE_CREDITS = 3
SYSTEM_PROMPT_SURCHARGE_CREDITS = 1
SYSTEM_PROMPT_FREE_CHARS = 500
RUN_STEP_CREDITS = 5
RUN_STEP_V1_CREDITS = 8
SNAPSHOT_CREDITS = 1
MACHINE_RUNNING_LINUX_CREDITS_PER_HOUR = 5
MACHINE_RUNNING_WINDOWS_CREDITS_PER_HOUR = 9
MACHINE_STOPPED_CREDITS_PER_HOUR = 1

_RUNNING_STATES = frozenset({"running", "starting", "stopping", "restarting"})
_STOPPED_STATES = frozenset({"stopped", "suspended", "suspended_for_billing"})
_FREE_STATES = frozenset({"creating", "error", "terminated"})


@dataclass(frozen=True)
class CostItem:
    """One itemized line of an estimate."""

    label: str
    credits: int


@dataclass(frozen=True)
class CostEstimate:
    """An itemized credit estimate. ``usd`` assumes a live (billing) key."""

    items: tuple[CostItem, ...]

    @property
    def credits(self) -> int:
        return sum(item.credits for item in self.items)

    @property
    def usd(self) -> float:
        return self.credits * CREDIT_USD


def is_hd(width: int, height: int) -> bool:
    """HD is strictly width > 1280 OR height > 720 (1280x720 itself is SD)."""
    return width > 1280 or height > 720


def _inference_surcharges(
    *,
    cua_version: CuaVersion,
    screen_width: int,
    screen_height: int,
    trajectory_sizes: Sequence[tuple[int, int]],
    system_prompt: str | None,
) -> list[CostItem]:
    items: list[CostItem] = []
    trajectory_count = len(trajectory_sizes)
    if trajectory_count:
        items.append(
            CostItem(
                f"trajectory screenshots x{trajectory_count} "
                f"(+{TRAJECTORY_SURCHARGE_CREDITS} cr each)",
                TRAJECTORY_SURCHARGE_CREDITS * trajectory_count,
            )
        )
    hd_count = int(is_hd(screen_width, screen_height)) + sum(
        1 for (width, height) in trajectory_sizes if is_hd(width, height)
    )
    if hd_count:
        items.append(
            CostItem(
                f"HD images x{hd_count} (+{HD_SURCHARGE_CREDITS} cr each, w>1280 or h>720)",
                HD_SURCHARGE_CREDITS * hd_count,
            )
        )
    if cua_version == "v1":
        items.append(CostItem("v1 engine surcharge", V1_SURCHARGE_CREDITS))
    if system_prompt is not None and len(system_prompt) > SYSTEM_PROMPT_FREE_CHARS:
        items.append(
            CostItem(
                f"system_prompt > {SYSTEM_PROMPT_FREE_CHARS} chars",
                SYSTEM_PROMPT_SURCHARGE_CREDITS,
            )
        )
    return items


def estimate_predict(
    *,
    cua_version: CuaVersion = "v3",
    screen_width: int = 1920,
    screen_height: int = 1080,
    trajectory_sizes: Sequence[tuple[int, int]] = (),
    system_prompt: str | None = None,
) -> CostEstimate:
    """Estimate one ``POST /v1/predict`` call (base 5 cr + surcharges)."""
    items = [CostItem("predict base", PREDICT_BASE_CREDITS)]
    items.extend(
        _inference_surcharges(
            cua_version=cua_version,
            screen_width=screen_width,
            screen_height=screen_height,
            trajectory_sizes=trajectory_sizes,
            system_prompt=system_prompt,
        )
    )
    return CostEstimate(items=tuple(items))


def estimate_session_create() -> CostEstimate:
    """Estimate ``POST /v1/sessions`` (flat 10 cr, no surcharges)."""
    return CostEstimate(items=(CostItem("session create", SESSION_CREATE_CREDITS),))


def estimate_session_predict(
    *,
    cua_version: CuaVersion = "v3",
    screen_width: int = 1920,
    screen_height: int = 1080,
    trajectory_sizes: Sequence[tuple[int, int]] = (),
    system_prompt: str | None = None,
) -> CostEstimate:
    """Estimate one session step (base 4 cr + the same surcharges as predict).

    ``trajectory_sizes`` are the sizes of the server-kept trajectory
    screenshots attached to this step.
    """
    items = [CostItem("session predict base", SESSION_PREDICT_BASE_CREDITS)]
    items.extend(
        _inference_surcharges(
            cua_version=cua_version,
            screen_width=screen_width,
            screen_height=screen_height,
            trajectory_sizes=trajectory_sizes,
            system_prompt=system_prompt,
        )
    )
    return CostEstimate(items=tuple(items))


def estimate_ground(*, screen_width: int = 1920, screen_height: int = 1080) -> CostEstimate:
    """Estimate ``POST /v1/ground`` (3 cr, +1 if the screenshot is HD)."""
    items = [CostItem("ground base", GROUND_BASE_CREDITS)]
    if is_hd(screen_width, screen_height):
        items.append(CostItem("HD image surcharge", HD_SURCHARGE_CREDITS))
    return CostEstimate(items=tuple(items))


def estimate_parse() -> CostEstimate:
    """``POST /v1/parse`` is free (deterministic, no model call)."""
    return CostEstimate(items=(CostItem("parse (free)", PARSE_CREDITS),))


def run_step_credits(cua_version: CuaVersion = "v3") -> int:
    """Credits for one completed run step: 8 on v1, 5 on v3/v4."""
    return RUN_STEP_V1_CREDITS if cua_version == "v1" else RUN_STEP_CREDITS


def estimate_run(*, steps: int, cua_version: CuaVersion = "v3") -> CostEstimate:
    """Estimate a task run of ``steps`` completed agent steps."""
    if steps < 0:
        raise ValueError("steps must be >= 0")
    per_step = run_step_credits(cua_version)
    return CostEstimate(
        items=(
            CostItem(f"run steps x{steps} ({per_step} cr each, {cua_version})", per_step * steps),
        )
    )


def estimate_workflow_run(*, task_steps: int, cua_version: CuaVersion = "v3") -> CostEstimate:
    """Estimate a workflow run: only ``task`` steps bill; control flow is free."""
    if task_steps < 0:
        raise ValueError("task_steps must be >= 0")
    per_step = run_step_credits(cua_version)
    return CostEstimate(
        items=(
            CostItem(
                f"workflow task steps x{task_steps} ({per_step} cr each, {cua_version})",
                per_step * task_steps,
            ),
            CostItem("control-flow steps (free)", 0),
        )
    )


def machine_hourly_credits(os_type: OsType, state: str) -> int:
    """Hourly runtime rate for a machine in the given state."""
    if state in _RUNNING_STATES:
        return (
            MACHINE_RUNNING_WINDOWS_CREDITS_PER_HOUR
            if os_type == "windows"
            else MACHINE_RUNNING_LINUX_CREDITS_PER_HOUR
        )
    if state in _STOPPED_STATES:
        return MACHINE_STOPPED_CREDITS_PER_HOUR
    if state in _FREE_STATES:
        return 0
    raise ValueError(f"unknown machine state: {state!r}")


def estimate_machine_runtime(
    *,
    os_type: OsType,
    state: str,
    minutes: float,
) -> CostEstimate:
    """Estimate machine runtime: hourly rate, metered per minute, rounded down."""
    if minutes < 0:
        raise ValueError("minutes must be >= 0")
    rate = machine_hourly_credits(os_type, state)
    credits = int(rate * minutes / 60)  # rounded down in your favor
    return CostEstimate(
        items=(
            CostItem(
                f"machine {os_type}/{state} x{minutes:g} min ({rate} cr/hr, rounded down)",
                credits,
            ),
        )
    )


def estimate_snapshot() -> CostEstimate:
    """``POST /machines/{id}/snapshot`` -- 1 cr one-time (refunded on failure)."""
    return CostEstimate(items=(CostItem("snapshot", SNAPSHOT_CREDITS),))


def combine(*estimates: CostEstimate) -> CostEstimate:
    """Merge several estimates into one itemized estimate."""
    items: list[CostItem] = []
    for estimate in estimates:
        items.extend(estimate.items)
    return CostEstimate(items=tuple(items))


def format_estimate(
    estimate: CostEstimate,
    *,
    title: str = "Estimated cost",
    sandbox: bool = False,
) -> str:
    """Render an itemized, printable estimate for examples to show pre-spend.

    With ``sandbox=True`` the dollar total is labeled $0.00 -- sandbox keys
    (``sk-coasty-test-*``) never bill.
    """
    lines = [f"{title}:"]
    for item in estimate.items:
        lines.append(f"  - {item.label}: {item.credits} cr")
    if sandbox:
        lines.append(f"  total: {estimate.credits} cr = $0.00 (sandbox key - never billed)")
    else:
        lines.append(f"  total: {estimate.credits} cr = ${estimate.usd:.2f}")
    return "\n".join(lines)

File: typescript/src/coasty/cost.ts
/**
 * Cost estimator implementing the full documented pricing table
 * (1 credit = 1 cent = $0.01 exactly). Sandbox keys (`sk-coasty-test-*`)
 * never bill — examples label those estimates "$0 (sandbox)".
 */
import { type CuaVersion, type MachineStatus, type OsType } from './types.js';

/** Full per-item pricing table, in credits. */
export const PRICING = {
  predict: 5,
  sessionCreate: 10, // one-time, no surcharges
  sessionPredict: 4,
  ground: 3,
  parse: 0,
  /** Surcharge: per trajectory screenshot attached. */
  trajectoryScreenshot: 2,
  /** Surcharge: per HD image (strictly width > 1280 OR height > 720). */
  hdImage: 1,
  /** Surcharge: per request on the v1 engine. */
  v1Engine: 3,
  /** Surcharge: system_prompt strictly longer than 500 chars (exactly 500 is free). */
  longSystemPrompt: 1,
  /** Run / workflow-task step on v3/v4 (no other surcharges apply to run steps). */
  runStepV3V4: 5,
  /** Run / workflow-task step on v1 (5 base + 3 engine surcharge). */
  runStepV1: 8,
  snapshot: 1,
  machineHourly: {
    linuxRunning: 5,
    windowsRunning: 9,
    /** Stopped / suspended (any OS): storage-only rate. */
    stopped: 1,
    /** Creating / error / terminated: free. */
    unbilled: 0,
  },
  /** Provisioning gate (NOT a fee): wallet must hold >= 20 credits. */
  provisioningGateCredits: 20,
} as const;

export const HD_WIDTH_THRESHOLD = 1280;
export const HD_HEIGHT_THRESHOLD = 720;
/** System prompts strictly longer than this add 1 credit. */
export const SYSTEM_PROMPT_FREE_CHARS = 500;

/** HD is STRICT: exactly 1280x720 is NOT HD. */
export function isHdImage(width: number, height: number): boolean {
  return width > HD_WIDTH_THRESHOLD || height > HD_HEIGHT_THRESHOLD;
}

export interface InferenceCostOptions {
  cuaVersion?: CuaVersion;
  /** Width of the screenshot you SEND (defaults mirror the API: 1920x1080 — note that IS HD). */
  screenWidth?: number;
  screenHeight?: number;
  /** Trajectory screenshots attached (assumed same dimensions as the current shot). */
  trajectoryScreenshots?: number;
  /** Length of `system_prompt` in characters (exactly 500 is still free). */
  systemPromptChars?: number;
}

function inferenceSurcharges(options: InferenceCostOptions): number {
  const cuaVersion = options.cuaVersion ?? 'v3';
  const width = options.screenWidth ?? 1920;
  const height = options.screenHeight ?? 1080;
  const trajectory = options.trajectoryScreenshots ?? 0;
  const promptChars = options.systemPromptChars ?? 0;

  let credits = trajectory * PRICING.trajectoryScreenshot;
  if (isHdImage(width, height)) {
    // The HD fee applies to the current screenshot AND each trajectory shot.
    credits += PRICING.hdImage * (1 + trajectory);
  }
  if (cuaVersion === 'v1') credits += PRICING.v1Engine;
  if (promptChars > SYSTEM_PROMPT_FREE_CHARS) credits += PRICING.longSystemPrompt;
  return credits;
}

/** POST /v1/predict: 5 credits + surcharges. */
export function estimatePredictCredits(options: InferenceCostOptions = {}): number {
  return PRICING.predict + inferenceSurcharges(options);
}

/** POST /v1/sessions: 10 credits one-time, NO surcharges. */
export function estimateSessionCreateCredits(): number {
  return PRICING.sessionCreate;
}

/** POST /v1/sessions/{id}/predict: 4 credits + the same surcharges as /predict. */
export function estimateSessionPredictCredits(options: InferenceCostOptions = {}): number {
  return PRICING.sessionPredict + inferenceSurcharges(options);
}

/** POST /v1/ground: 3 credits, +1 if the screenshot is HD. */
export function estimateGroundCredits(
  options: { screenWidth?: number; screenHeight?: number } = {},
): number {
  const width = options.screenWidth ?? 1920;
  const height = options.screenHeight ?? 1080;
  return PRICING.ground + (isHdImage(width, height) ? PRICING.hdImage : 0);
}

/** POST /v1/parse is free. */
export function estimateParseCredits(): number {
  return PRICING.parse;
}

/** Per-step run price: 5 credits on v3/v4, 8 on v1. No other surcharges on run steps. */
export function runStepCredits(cuaVersion: CuaVersion = 'v3'): number {
  return cuaVersion === 'v1' ? PRICING.runStepV1 : PRICING.runStepV3V4;
}

/** Estimate a whole run: steps × per-step rate. */
export function estimateRunCredits(options: { steps: number; cuaVersion?: CuaVersion }): number {
  return options.steps * runStepCredits(options.cuaVersion ?? 'v3');
}

/**
 * Estimate a workflow run. Only `task` steps bill (at the run-step rate ×
 * their expected agent steps); control-flow steps are free.
 */
export function estimateWorkflowRunCredits(options: {
  /** Total agent steps expected across all `task` steps. */
  taskSteps: number;
  cuaVersion?: CuaVersion;
}): number {
  return options.taskSteps * runStepCredits(options.cuaVersion ?? 'v3');
}

const RUNNING_STATES: ReadonlySet<MachineStatus> = new Set([
  'running',
  'starting',
  'stopping',
  'restarting',
]);
const STOPPED_STATES: ReadonlySet<MachineStatus> = new Set(['stopped', 'suspended_for_billing']);

/**
 * Hourly runtime rate by OS and state. Running (incl. starting/stopping/
 * restarting): Linux 5, Windows 9. Stopped/suspended (any OS): 1.
 * Creating/error/terminated: 0.
 */
export function machineHourlyCredits(osType: OsType, status: MachineStatus): number {
  if (RUNNING_STATES.has(status)) {
    return osType === 'windows'
      ? PRICING.machineHourly.windowsRunning
      : PRICING.machineHourly.linuxRunning;
  }
  if (STOPPED_STATES.has(status)) return PRICING.machineHourly.stopped;
  return PRICING.machineHourly.unbilled;
}

/**
 * Runtime is metered per minute and billed in whole credits, ROUNDED DOWN in
 * your favor (partial credits are never billed).
 */
export function estimateMachineRuntimeCredits(options: {
  osType: OsType;
  minutes: number;
  status?: MachineStatus;
}): number {
  const hourly = machineHourlyCredits(options.osType, options.status ?? 'running');
  return Math.floor((hourly * options.minutes) / 60);
}

/** POST /v1/machines/{id}/snapshot: 1 credit one-time (refunded on failure). */
export function snapshotCredits(): number {
  return PRICING.snapshot;
}

/** 1 credit = 1 cent = $0.01, exactly. */
export function creditsToUsd(credits: number): number {
  return credits / 100;
}

export function formatUsd(credits: number): string {
  return `$${creditsToUsd(credits).toFixed(2)}`;
}

export interface EstimateLineItem {
  label: string;
  credits: number;
}

/**
 * Render an itemized cost estimate for examples to print before spending.
 * When `sandbox` is true the total is labelled `$0.00 (sandbox key — never bills)`.
 */
export function formatEstimate(
  items: readonly EstimateLineItem[],
  options: { sandbox?: boolean } = {},
): string {
  const total = items.reduce((sum, item) => sum + item.credits, 0);
  const labelWidth = Math.max(5, ...items.map((item) => item.label.length));
  const lines = items.map(
    (item) =>
      `  ${item.label.padEnd(labelWidth)}  ${String(item.credits).padStart(6)} cr  (${formatUsd(item.credits)})`,
  );
  const totalText =
    options.sandbox === true
      ? `  ${'TOTAL'.padEnd(labelWidth)}  ${String(total).padStart(6)} cr  ($0.00 — sandbox key, never bills)`
      : `  ${'TOTAL'.padEnd(labelWidth)}  ${String(total).padStart(6)} cr  (${formatUsd(total)})`;
  return ['Estimated cost:', ...lines, totalText].join('\n');
}
